bittide-shake/src/Clash/Shake/Cargo.hs

```diff
@@ -0,0 +1,31 @@
+-- SPDX-FileCopyrightText: 2023 Google LLC
+--
+-- SPDX-License-Identifier: Apache-2.0
+module Clash.Shake.Cargo where
+
+import Prelude
+
+import System.Exit
+import System.IO
+import System.Process
+
+cargoBuildFirmwareProgram ::
+  FilePath ->
+  -- ^ Directory to perform the call in
+  Bool ->
+  -- ^ release mode?
+  IO ()
+cargoBuildFirmwareProgram workingDir release = do
+  let
+    createProc =
+      (proc "cargo" (["build"] <> ["--release" | release]))
+        { cwd = Just workingDir}
+  (exitCode, _stdoutOut, stderrOut) <- readCreateProcessWithExitCode createProc ""
+
+  case exitCode of
+    ExitSuccess -> do
+      pure ()
+    ExitFailure _ -> do
+      hPutStrLn stderr "Compilation failed"
```

Comment:
Could you include what command exactly failed? E.g., when something failed I'd much rather see:

```haskell
{-# LANUGAGE QuasiQuotes #-}

import qualified Clash.Util.Interpolate as I

[..]
  case exitCode of
    ExitSuccess -> do
      pure ()
    ExitFailure _ -> do
      hPutStrLn stderr [I.i|
        `#{cmd} #{args}` terminated with exit code `3`. Stderr:

          #{stderr}

        Stdout:

          #{stdout}
     |]
```

It's slightly grimy to use `Clash.Util.Interpolate`, but I don't think that's going away any time soon :upside_down_face:. (Also note that the quasiquoter will correctly indent if stdout/stderr is a multiline string.)